src/finn/custom_op/fpgadataflow/thresholding_binary_search.py

```diff
@@ -0,0 +1,752 @@
+            )
+        # ensure channels = mh , duplicating if necessary
+        if ret.shape[0] == 1:
+            ret = np.tile(ret, (mh, 1))
+        assert (
+            ret.shape[0] == mh
+        ), "Channels of threshold matrix are not as expected (mh)"
+        # distribute rows between PEs
+        ret = interleave_matrix_outer_dim_from_partitions(ret, pe)
+        assert (
+            ret.shape[0] == pe
+        ), """First dimension after distribution of the
+        rows between PEs is not as expected (pe)"""
+        assert (
+            ret.shape[1] == tmem
+        ), """Second dimension after distribution of the
+        rows between PEs is not as expected (tmem)"""
+        assert (
+            ret.shape[2] == n_thres_steps
+        ), """Third dimension after distribution of the
+        rows between PEs is not as expected (n_thres_steps)"""
+        return ret.reshape(1, pe, tmem, n_thres_steps)
+
+    def make_weight_file(self, weights, weight_file_mode, weight_file_name):
+        """Produce a file containing given weights (thresholds) in appropriate
+        format for this layer. This file can be used for either synthesis or
+        run-time reconfig of weights.
+
+        Arguments:
+        * weights : numpy array with weights to be put into the file
+        * weight_file_mode : one of {hls_header, decoupled_verilog_dat,
+          decoupled_runtime}
+        * weight_file_name : filename for the weight file to be generated
+        """
+        # There are 'decoupled_*' flavors, just make sure that the flavors
+        # are decoupled related
+        if "decoupled" not in weight_file_mode:
+            raise Exception(
+                "Unrecognized memory mode for this node: {}".format(weight_file_mode)
+            )
+
+        threshold_tensor = self.get_hls_compatible_threshold_tensor(weights)
+        tdt = self.get_weight_datatype()
+        assert np.vectorize(tdt.allowed)(
+            threshold_tensor
+        ).all(), "Thresholds can't be expressed with type %s" % str(tdt)
+
+        # streaming thresholds need to be organized differently
+        # (1, pe, tmem, n_thres_steps) -> (1, tmem, pe, n_thres_steps)
+        decoupled_thres = np.transpose(threshold_tensor, (0, 2, 1, 3))
+        # (1, tmem, pe, n_thres_steps) -(1, tmem, pe * n_thres_steps)
+        pe = self.get_nodeattr("PE")
+        n_thres_steps = self.get_nodeattr("numSteps")
+        decoupled_thres_pe_flipped = np.flip(decoupled_thres, axis=-2)
+        decoupled_thres = decoupled_thres.reshape(1, -1, pe * n_thres_steps)
+        decoupled_thres = decoupled_thres.copy()
+        decoupled_thres_pe_flipped = decoupled_thres_pe_flipped.reshape(
+            1, -1, pe * n_thres_steps
+        )
+        decoupled_thres_pe_flipped = decoupled_thres_pe_flipped.copy()
+
+        if weight_file_mode == "decoupled_npy":
+            # save weight stream into npy for cppsim
+            np.save(weight_file_name, decoupled_thres)
```

Comment:
there is no cppsim for RTL components

Reply:
removed in https://github.com/Xilinx/finn/pull/687/commits/49bdd28e4edc3d47ccb57161e073fcde2a2cb216